Update Roc

JanCVanB opened the pull request:
Roc changes fast, so updates like this will be necessary from time to time.

JanCVanB commented:
I was getting error A, but after adding `freetype = "0.7.0"`, I get error B instead.

## Error A

```
[jan@framey platform]$ (cd ../roc && cargo run ../examples/shapes.roc)
    Finished dev [unoptimized + debuginfo] target(s) in 1.63s
     Running `target/debug/roc ../examples/shapes.roc`
🔨 Rebuilding host...
thread '<unnamed>' panicked at 'Failed to rebuild src/main.rs - stderr of the `cargo build` command was:
   Compiling host v0.1.0 (/home/jan/_code/_roc/roc-plotters/platform)
   Compiling servo-fontconfig-sys v5.1.0
error: failed to run custom build command for `servo-fontconfig-sys v5.1.0`

Caused by:
  process didn't exit successfully: `/home/jan/_code/_roc/roc-plotters/platform/target/debug/build/servo-fontconfig-sys-1a51cdc3c07f13df/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=FONTCONFIG_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=FONTCONFIG_STATIC
  cargo:rerun-if-env-changed=FONTCONFIG_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cd /home/jan/_code/_roc/roc-plotters/platform/target/debug/build/servo-fontconfig-sys-4bcdbcf382354214/out && \
  	CC="gcc" \
  	AR="ar" \
  	FREETYPE_CFLAGS="" \
  	FREETYPE_LIBS="" \
  	CFLAGS=""" -fPIC" \
  	/home/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/servo-fontconfig-sys-5.1.0/configure \
  	--disable-docs \
  	--disable-shared \
  	\
  	--host=x86_64-unknown-linux-gnu --sysconfdir=/etc --localstatedir=/var && \
  	touch /home/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/servo-fontconfig-sys-5.1.0/src/fcobjshash.h
  checking for a BSD-compatible install... /nix/store/7jr7pr4c6yb85xpzay5xafs5zlcadkhz-coreutils-9.0/bin/install -c
  checking whether build environment is sane... yes
  checking for x86_64-unknown-linux-gnu-strip... strip
  checking for a thread-safe mkdir -p... /nix/store/7jr7pr4c6yb85xpzay5xafs5zlcadkhz-coreutils-9.0/bin/mkdir -p
  checking for gawk... gawk
  checking whether make sets $(MAKE)... yes
  checking whether make supports nested variables... yes
  checking whether make supports nested variables... (cached) yes
  checking whether to enable maintainer-specific portions of Makefiles... no
  checking for x86_64-unknown-linux-gnu-gcc... gcc
  checking whether the C compiler works... yes
  checking for C compiler default output file name... a.out
  checking for suffix of executables... 
  checking whether we are cross compiling... no
  checking for suffix of object files... o
  checking whether we are using the GNU C compiler... yes
  checking whether gcc accepts -g... yes
  checking for gcc option to accept ISO C89... none needed
  checking whether gcc understands -c and -o together... yes
  checking for style of include used by make... GNU
  checking dependency style of gcc... gcc3
  checking how to run the C preprocessor... gcc -E
  checking for grep that handles long lines and -e... /nix/store/qd9jxc0q00cr7fp30y6jbbww20gj33lg-gnugrep-3.7/bin/grep
  checking for egrep... /nix/store/qd9jxc0q00cr7fp30y6jbbww20gj33lg-gnugrep-3.7/bin/grep -E
  checking for ANSI C header files... yes
  checking for sys/types.h... yes
  checking for sys/stat.h... yes
  checking for stdlib.h... yes
  checking for string.h... yes
  checking for memory.h... yes
  checking for strings.h... yes
  checking for inttypes.h... yes
  checking for stdint.h... yes
  checking for unistd.h... yes
  checking minix/config.h usability... no
  checking minix/config.h presence... no
  checking for minix/config.h... no
  checking whether it is safe to define __EXTENSIONS__... yes
  checking for special C compiler options needed for large files... no
  checking for _FILE_OFFSET_BITS value needed for large files... no
  checking whether ln -s works... yes
  checking whether make sets $(MAKE)... (cached) yes
  checking for x86_64-unknown-linux-gnu-pkg-config... no
  checking for pkg-config... /nix/store/szc9mx70l0lk32rbyixpq5mska4fzj61-pkg-config-wrapper-0.29.2/bin/pkg-config
  checking pkg-config is at least version 0.9.0... yes
  checking for RM macro... rm -f
  checking build system type... x86_64-unknown-linux-gnu
  checking host system type... x86_64-unknown-linux-gnu
  checking how to print strings... printf
  checking for a sed that does not truncate output... /nix/store/lgvd2fh4cndlv8mnyy49jp1nplpml3xp-gnused-4.8/bin/sed
  checking for fgrep... /nix/store/qd9jxc0q00cr7fp30y6jbbww20gj33lg-gnugrep-3.7/bin/grep -F
  checking for ld used by gcc... ld
  checking if the linker (ld) is GNU ld... yes
  checking for BSD- or MS-compatible name lister (nm)... nm
  checking the name lister (nm) interface... BSD nm
  checking the maximum length of command line arguments... 1572864
  checking whether the shell understands some XSI constructs... yes
  checking whether the shell understands "+="... yes
  checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
  checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
  checking for ld option to reload object files... -r
  checking for x86_64-unknown-linux-gnu-objdump... objdump
  checking how to recognize dependent libraries... pass_all
  checking for x86_64-unknown-linux-gnu-dlltool... no
  checking for dlltool... no
  checking how to associate runtime and link libraries... printf %s\n
  checking for x86_64-unknown-linux-gnu-ar... ar
  checking for archiver @FILE support... @
  checking for x86_64-unknown-linux-gnu-strip... (cached) strip
  checking for x86_64-unknown-linux-gnu-ranlib... ranlib
  checking command to parse nm output from gcc object... ok
  checking for sysroot... no
  checking for x86_64-unknown-linux-gnu-mt... no
  checking for mt... no
  checking if : is a manifest tool... no
  checking for dlfcn.h... yes
  checking for objdir... .libs
  checking if gcc supports -fno-rtti -fno-exceptions... no
  checking for gcc option to produce PIC... -fPIC -DPIC
  checking if gcc PIC flag -fPIC -DPIC works... yes
  checking if gcc static flag -static works... no
  checking if gcc supports -c -o file.o... yes
  checking if gcc supports -c -o file.o... (cached) yes
  checking whether the gcc linker (ld -m elf_x86_64) supports shared libraries... yes
  checking dynamic linker characteristics... cflags:  -fPIC
  cppflags: 
  ldflags:  -Wl,-rpath -Wl,/foo
  libs: 
  GNU/Linux ld.so
  checking how to hardcode library paths into programs... immediate
  checking whether stripping libraries is possible... yes
  checking if libtool supports shared libraries... yes
  checking whether to build shared libraries... no
  checking whether to build static libraries... yes
  checking for dirent.h that defines DIR... yes
  checking for library containing opendir... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  none required
  checking for ANSI C header files... (cached) yes
  checking fcntl.h usability... yes
  checking fcntl.h presence... yes
  checking for fcntl.h... yes
  checking regex.h usability... yes
  checking regex.h presence... yes
  checking for regex.h... yes
  checking for stdlib.h... (cached) yes
  checking for string.h... (cached) yes
  checking for unistd.h... (cached) yes
  checking sys/statvfs.h usability... yes
  checking sys/statvfs.h presence... yes
  checking for sys/statvfs.h... yes
  checking sys/vfs.h usability... yes
  checking sys/vfs.h presence... yes
  checking for sys/vfs.h... yes
  checking sys/statfs.h usability... yes
  checking sys/statfs.h presence... yes
  checking for sys/statfs.h... yes
  checking sys/param.h usability... yes
  checking sys/param.h presence... yes
  checking for sys/param.h... yes
  checking sys/mount.h usability... yes
  checking sys/mount.h presence... yes
  checking for sys/mount.h... yes
  checking for stdint types... stdint.h (shortcircuit)
  make use of stdint.h in src/fcstdint.h (assuming C99 compatible system)
  checking for an ANSI C-conforming const... yes
  checking for inline... inline
  checking for flexible array members... yes
  checking for pid_t... yes
  checking for vprintf... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  conftest.c:65:6: warning: conflicting types for built-in function 'vprintf'; expected 'int(const char *, __va_list_tag *)' [-Wbuiltin-declaration-mismatch]
     65 | char vprintf ();
        |      ^~~~~~~
  conftest.c:53:1: note: 'vprintf' is declared in header '<stdio.h>'
     52 | # include <limits.h>
    +++ |+#include <stdio.h>
     53 | #else
  yes
  checking for _doprnt... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  /nix/store/rq6bh3qfrqnyqwik0w3q6w180zg3w2pa-binutils-2.38/bin/ld: /tmp/nix-shell.40gdPb/ccW6aqJx.o: in function `main':
  conftest.c:(.text.startup+0x7): undefined reference to `_doprnt'
  collect2: error: ld returned 1 exit status
  no
  checking for stdlib.h... (cached) yes
  checking for unistd.h... (cached) yes
  checking for sys/param.h... (cached) yes
  checking for getpagesize... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for working mmap... yes
  checking for link... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for mkstemp... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for mkostemp... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for _mktemp_s... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  /nix/store/rq6bh3qfrqnyqwik0w3q6w180zg3w2pa-binutils-2.38/bin/ld: /tmp/nix-shell.40gdPb/ccDbzrvF.o: in function `main':
  conftest.c:(.text.startup+0x7): undefined reference to `_mktemp_s'
  collect2: error: ld returned 1 exit status
  no
  checking for mkdtemp... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for getopt... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for getopt_long... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for getprogname... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  /nix/store/rq6bh3qfrqnyqwik0w3q6w180zg3w2pa-binutils-2.38/bin/ld: /tmp/nix-shell.40gdPb/ccNML1Hh.o: in function `main':
  conftest.c:(.text.startup+0x7): undefined reference to `getprogname'
  collect2: error: ld returned 1 exit status
  no
  checking for getexecname... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  /nix/store/rq6bh3qfrqnyqwik0w3q6w180zg3w2pa-binutils-2.38/bin/ld: /tmp/nix-shell.40gdPb/cc9iC9oq.o: in function `main':
  conftest.c:(.text.startup+0x7): undefined reference to `getexecname'
  collect2: error: ld returned 1 exit status
  no
  checking for rand... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for random... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for lrand48... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for random_r... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for rand_r... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for readlink... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for regcomp... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for regerror... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for regexec... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for regfree... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for fstatvfs... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for fstatfs... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for lstat... cflags:  -fPIC
  cppflags: 
  ldflags: 
  libs: 
  yes
  checking for posix_fadvise in fcntl.h... fcntl.h
  checking for scandir... yes
  checking for struct statvfs.f_basetype... no
  checking for struct statvfs.f_fstypename... no
  checking for struct statfs.f_flags... yes
  checking for struct statfs.f_fstypename... no
  checking for struct dirent.d_type... yes
  checking for FREETYPE... no

  --- stderr
  configure: error: Package requirements (freetype2) were not met:

  No package 'freetype2' found

  Consider adjusting the PKG_CONFIG_PATH environment variable if you
  installed software in a non-standard prefix.

  Alternatively, you may set the environment variables FREETYPE_CFLAGS
  and FREETYPE_LIBS to avoid the need to call pkg-config.
  See the pkg-config man page for more details.
  make: *** [makefile.cargo:83: /home/jan/_code/_roc/roc-plotters/platform/target/debug/build/servo-fontconfig-sys-4bcdbcf382354214/out/Makefile] Error 1
  thread 'main' panicked at 'assertion failed: Command::new(\"make\").env(\"MAKEFLAGS\",\n                    env::var(\"CARGO_MAKEFLAGS\").unwrap_or_default()).args(&[\"-R\",\n                            \"-f\", \"makefile.cargo\"]).status().unwrap().success()', /home/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/servo-fontconfig-sys-5.1.0/build.rs:41:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
', crates/compiler/build/src/link.rs:1199:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', crates/cli/src/build.rs:283:46
[jan@framey platform]$ 
```

## Error B

```
[jan@framey roc]$ cargo run ../examples/hello_world.roc 
    Finished dev [unoptimized + debuginfo] target(s) in 0.47s
     Running `target/debug/roc ../examples/hello_world.roc`
🔨 Rebuilding host...
thread '<unnamed>' panicked at 'Failed to rebuild src/main.rs - stderr of the `cargo build` command was:
   Compiling host v0.1.0 (/home/jan/_code/_roc/roc-plotters/platform)
   Compiling yeslogic-fontconfig-sys v3.2.0
   Compiling semver v0.11.0
   Compiling num-traits v0.2.15
   Compiling freetype-sys v0.13.1
   Compiling dirs-sys-next v0.1.2
   Compiling time v0.1.44
error: failed to run custom build command for `yeslogic-fontconfig-sys v3.2.0`

Caused by:
  process didn't exit successfully: `/home/jan/_code/_roc/roc-plotters/platform/target/debug/build/yeslogic-fontconfig-sys-23a8645d0db93cda/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=RUST_FONTCONFIG_DLOPEN
  cargo:rerun-if-env-changed=FONTCONFIG_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=FONTCONFIG_STATIC
  cargo:rerun-if-env-changed=FONTCONFIG_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR

  --- stderr
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "`\"pkg-config\" \"--libs\" \"--cflags\" \"fontconfig\"` did not exit successfully: exit status: 1\nerror: could not find system library 'fontconfig' required by the 'yeslogic-fontconfig-sys' crate\n\n--- stderr\nPackage fontconfig was not found in the pkg-config search path.\nPerhaps you should add the directory containing `fontconfig.pc'\nto the PKG_CONFIG_PATH environment variable\nNo package 'fontconfig' found\n"', /home/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/yeslogic-fontconfig-sys-3.2.0/build.rs:8:48
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
', crates/compiler/build/src/link.rs:1199:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', crates/cli/src/build.rs:283:46
[jan@framey roc]$ git status
On branch trunk
Your branch is up to date with 'origin/trunk'.

nothing to commit, working tree clean
[jan@framey roc]$ 
```

Anton-4 commented:
I replaced shell.nix with a flake to make it easy to add dependencies necessary for the font-related errors.
Until we release roc on nix you do still have to update the path of roc in the nix flake, I've updated the instructions in the README.

JanCVanB commented:
Whoa! I never would have figured this out, thank you very much. What do you mean by "release roc on nix"?

Anton-4 commented:
As in roc will be [a package on nix](https://search.nixos.org/packages) :)
I'm working on that this week.

Anton-4 commented:
btw I will file issues on the roc repo for the math and scatter errors.

Anton-4 commented:
For the math issue: https://github.com/rtfeldman/roc/issues/3609

Anton-4 commented:
For the scatter issue: https://github.com/rtfeldman/roc/issues/3610

JanCVanB commented:
Woohoo! Thank you again 😃

Anton-4 commented:
I was able to fix the scatter example, the math example issue needs to be fixed in the compiler.

JanCVanB commented:
@Anton-4 Am I doing something silly?
```
[jan@framey roc-plotters]$ git status
On branch update_roc
Your branch is up to date with 'origin/update_roc'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   flake.nix

no changes added to commit (use "git add" and/or "git commit -a")
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ git diff
diff --git a/flake.nix b/flake.nix
index 381f717..23af44c 100644
--- a/flake.nix
+++ b/flake.nix
@@ -2,7 +2,7 @@
   description = "roc-plotters";
 
   inputs = {
-    roc.url = "path:/home/anton/gitrepos/roc-plotters/roc";
+    roc.url = "path:/home/jan/_code/_roc/roc-plotters/roc";
     # to easily make configs for multiple architectures
     flake-utils.url = "github:numtide/flake-utils";
   };
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ nix develop
warning: Git tree '/home/jan/_code/_roc/roc-plotters' is dirty
warning: updating lock file '/home/jan/_code/_roc/roc-plotters/flake.lock':
• Updated input 'roc':
    'path:/home/anton/gitrepos/roc-plotters/roc?lastModified=1658494956&narHash=sha256-8R0S4dSmIQjfwHwOBE3vKvCbIeCUr+TYzhrMvFdAtIE=' (2022-07-22)
  → 'path:/home/jan/_code/_roc/roc-plotters/roc?lastModified=1658467068&narHash=sha256-FeI4gEfmQqbBTsxQRfOeMNZU1LLV372gNcvJcp6LDuY=' (2022-07-22)
warning: Git tree '/home/jan/_code/_roc/roc-plotters' is dirty
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ cd roc
[jan@framey roc]$ 
[jan@framey roc]$ 
[jan@framey roc]$ cargo run ../examples/hello_world.roc
    Finished dev [unoptimized + debuginfo] target(s) in 1.23s
     Running `target/debug/roc ../examples/hello_world.roc`
🔨 Rebuilding host...
Fontconfig error: No writable cache directories
Illegal instruction (core dumped)
[jan@framey roc]$ 
```

Anton-4 commented:
I'll take a look on monday

Anton-4 commented:
I was not yet able to reproduce this. I've seen the `illegal instruction` error before when people were using a virtual machine, WSL(Windows Subsytem for Linux), or when they had a "special" CPU. Do any of those apply to you @JanCVanB?

I also don't understand how Fontconfig can error here, I'm clueless as to from where it would even be called.

JanCVanB commented:
I don't think my setup is strange in any of those ways, as I'm running a minimal Fedora 36 setup on a Framework laptop. Hmmm...

JanCVanB commented:
FWIW:
```
[jan@framey ~]$ neofetch
             .',;::::;,'.                jan@framey 
         .';:cccccccccccc:;,.            ---------- 
      .;cccccccccccccccccccccc;.         OS: Fedora Linux 36 (Workstation Edition) x86_64 
    .:cccccccccccccccccccccccccc:.       Host: Laptop AA 
  .;ccccccccccccc;.:dddl:.;ccccccc;.     Kernel: 5.18.10-200.fc36.x86_64 
 .:ccccccccccccc;OWMKOOXMWd;ccccccc:.    Uptime: 5 hours, 6 mins 
.:ccccccccccccc;KMMc;cc;xMMc:ccccccc:.   Packages: 1997 (rpm), 42 (nix-default), 15 (flatpak) 
,cccccccccccccc;MMM.;cc;;WW::cccccccc,   Shell: bash 5.1.16 
:cccccccccccccc;MMM.;cccccccccccccccc:   Resolution: 2256x1504 
:ccccccc;oxOOOo;MMM0OOk.;cccccccccccc:   DE: GNOME 42.3.1 
cccccc:0MMKxdd:;MMMkddc.;cccccccccccc;   WM: Mutter 
ccccc:XM0';cccc;MMM.;cccccccccccccccc'   WM Theme: Adwaita 
ccccc;MMo;ccccc;MMW.;ccccccccccccccc;    Theme: Adwaita [GTK2/3] 
ccccc;0MNc.ccc.xMMd:ccccccccccccccc;     Icons: Adwaita [GTK2/3] 
cccccc;dNMWXXXWM0::cccccccccccccc:,      Terminal: kitty 
cccccccc;.:odl:.;cccccccccccccc:,.       CPU: 11th Gen Intel i5-1135G7 (8) @ 4.200GHz 
:cccccccccccccccccccccccccccc:'.         GPU: Intel TigerLake-LP GT2 [Iris Xe Graphics] 
.:cccccccccccccccccccccc:;,..            Memory: 4439MiB / 15684MiB 
  '::cccccccccccccc::;,.
  ```

JanCVanB commented:
No, that seems to work fine:
```
[jan@framey ~]$ cd _code/_roc/roc-plotters/
[jan@framey roc-plotters]$ cd roc
[jan@framey roc]$ nix develop
[jan@framey roc]$ cargo run examples/hello-world/main.roc 
    Finished dev [unoptimized + debuginfo] target(s) in 1.03s
     Running `target/debug/roc examples/hello-world/main.roc`
🔨 Rebuilding host...
Hello, World!
[jan@framey roc-plotters]$ exit
exit
[jan@framey roc]$ cd ..
[jan@framey roc-plotters]$ nix develop
warning: Git tree '/home/jan/_code/_roc/roc-plotters' is dirty
[jan@framey roc-plotters]$ cd roc
[jan@framey roc]$ cargo run ../examples/hello_world.
hello_world.png  hello_world.roc  
[jan@framey roc]$ cargo run ../examples/hello_world.roc 
    Finished dev [unoptimized + debuginfo] target(s) in 0.47s
     Running `target/debug/roc ../examples/hello_world.roc`
🔨 Rebuilding host...
Fontconfig error: No writable cache directories
Illegal instruction (core dumped)
[jan@framey roc]$ 
```

Anton-4 commented:
Ok things make more sense now, I thought you were running the roc hello world example, but it's the one from roc-plotters.

Anton-4 commented:
I'll try reproducing it on my Ubuntu VM first and will set up a Fedora VM if necessary.

Anton-4 commented:
The `Fontconfig error ...` reproduces on Ubuntu, it is printed 138 times :p

Anton-4 commented:
I haven't been able to work around the `Fontconfig error` but I don't think it causes the `Illegal instruction`.
I'm downloading fedora now.

Anton-4 commented:
On my fedora VM I also have a bunch of Fontconfig errors but like on Ubuntu, the program completes successfully.
@JanCVanB can you share the output of `lscpu`?

Anton-4 commented:
Can you also share your hello_world binary from the roc-plotters/examples folder?
That way I can look for differences between mine and yours.

JanCVanB commented:
@Anton-4 
```
[jan@framey roc-plotters]$ lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-7
Vendor ID:               GenuineIntel
  Model name:            11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    CPU family:          6
    Model:               140
    Thread(s) per core:  2
    Core(s) per socket:  4
    Socket(s):           1
    Stepping:            1
    CPU(s) scaling MHz:  20%
    CPU max MHz:         4200.0000
    CPU min MHz:         400.0000
    BogoMIPS:            4838.40
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts ac
                         pi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_p
                         erfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pn
                         i pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_
                         1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 
                         3dnowprefetch cpuid_fault epb cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_e
                         nhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 sme
                         p bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb 
                         intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_d
                         etect dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512
                         vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512
                         _vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arc
                         h_capabilities
Virtualization features: 
  Virtualization:        VT-x
Caches (sum of all):     
  L1d:                   192 KiB (4 instances)
  L1i:                   128 KiB (4 instances)
  L2:                    5 MiB (4 instances)
  L3:                    8 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-7
Vulnerabilities:         
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Not affected
  Retbleed:              Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
  Srbds:                 Not affected
  Tsx async abort:       Not affected
[jan@framey roc-plotters]$ 
```

Anton-4 commented:
Thanks, nothing looks off there.
Can you send over the roc-plotters/examples/hello_world binary?

JanCVanB commented:
I'm not sure how, since GitHub apparently doesn't allow binary uploads in comments! I'll just commit it temporarily, haha

Anton-4 commented:
freetype is dynamically linked in my executable but not in yours, which is strange.
Can you remove everything in the platform folder that is not saved on git, like dynhost, libapp.so... and do `cargo run ../examples/hello_world.roc` again?

JanCVanB commented:
Sure! Thanks again for helping with this :)
```
[jan@framey roc-plotters]$ git status
On branch update_roc
Your branch is up to date with 'origin/update_roc'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   examples/hello_world
	modified:   flake.nix

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	platform/dynhost
	platform/libapp.so
	platform/metadata
	platform/preprocessedhost

no changes added to commit (use "git add" and/or "git commit -a")
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ rm examples/hello_world platform/dynhost platform/libapp.so platform/metadata platform/preprocessedhost 
[jan@framey roc-plotters]$ 
[jan@framey roc-plotters]$ cd roc
[jan@framey roc]$
[jan@framey roc]$ nix develop
[jan@framey roc]$
[jan@framey roc]$ cargo run ../examples/hello_world.roc 
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/roc ../examples/hello_world.roc`
🔨 Rebuilding host...
Fontconfig error: No writable cache directories
Illegal instruction (core dumped)
[jan@framey roc]$
```

Anton-4 commented:
I'll try to look at this again next week.

Anton-4 commented:
@JanCVanB does the illegal instruction error still happen?

JanCVanB commented:
I was getting different errors, even when running in a nic shell session that was started inside the Roc repo. With the `remove_roc` branch I'm contemplating removing the shell setup stuff from this repo entirely (and just relying on Roc & it's dependencies being installed already) but I don't know if that actually fits the platform development model. That seems like a more viable path for an app/library repo. I'll try this again another time on different computers and see what legitimate error messages I get now. Do you see any error messages? Thanks, @Anton-4 !

Anton-4 commented:
Starting a nix shell from inside the repo may lead you to missing freetype, expat, fontconfig...
I can reproduce the illegal instruction error now that I am on nixOS, I'll try to look at it when I'm waiting for CI builds :)
When updating roc you also want to delete flake.lock (in the roc-plotters folder), nix does not check if the roc folder has changed.

Anton-4 commented:
Hmm, some strange behavior; if I first do `./roc/target/release/roc examples/hello_world.roc` and then `./examples/hello_world`, no errors occur. Does the same happen for you?

Anton-4 commented:
I think you forgot to post the output @JanCVanB :)

Anton-4 commented:
Can you also tell me what just `./examples/hello_world` returns after having done one of the commands you listed?

Anton-4 commented:
friendly ping @JanCVanB :)

JanCVanB commented:
friendly pong! Haha yes thank you for reminding me about this

JanCVanB commented:
Here's my output that I forgot to post:
```
[jan@framey roc-plotters]$ cd roc && cargo run ../examples/hello_world.roc 
    Finished dev [unoptimized + debuginfo] target(s) in 0.42s
     Running `target/debug/roc ../examples/hello_world.roc`
🔨 Rebuilding platform...
Fontconfig error: No writable cache directories
Illegal instruction (core dumped)
[jan@framey roc]$ cd ..
[jan@framey roc-plotters]$ cd roc && cargo build --release && cd .. && ./roc/target/release/roc ./examples/hello_world.roc
    Finished release [optimized] target(s) in 0.27s
🔨 Rebuilding platform...
Fontconfig error: No writable cache directories
Illegal instruction (core dumped)
[jan@framey roc-plotters]$ 
```
(I only now notice that the latter command leaves the shell in a different directory than the former... probably unimportant)

JanCVanB commented:
and here's what you asked for:
```
[jan@framey roc-plotters]$ ./examples/hello_world 
Illegal instruction (core dumped)
[jan@framey roc-plotters]$ 
```

JanCVanB commented:
(I updated the submodule to the latest `main` commit before running those commands, btw... and I don't remember if these outputs are the same as I saw 2 weeks ago)

Anton-4 commented:
:+1: thanks!

JanCVanB commented:
@Anton-4 I'm considering abandoning this PR and just deleting the submodule & scripts. This platform shouldn't depend on the roc repo, right? We can just use roc from nightly build download?

JanCVanB commented:
Those 3 commits I just pushed are from git cloning on my mac x86 and running roc nightly on the examples... no submodule necessary, I think (though I did pull & update the submodule, so maybe it's being used)

Anton-4 commented:
> This platform shouldn't depend on the roc repo, right? We can just use roc from nightly build download?
Yep, that should work
